### PR TITLE
Only negate index expression on all indices with preceding wildcard

### PR DESCRIPTION
**Pull request by dakrone**
There is currently a very confusing behavior in Elasticsearch for the
following:

Given the indices: `[test1, test2, -foo1, -foo2]`

```
DELETE /-foo*
```

Will cause the `test1` and `test2` indices to be deleted, when what is
usually intended is to delete the `-foo1` and `-foo2` indices.

Previously we added a change in #20033 to disallow creating indices
starting with `-` or `+`, which will help with this situation. However,
users may have existing indices starting with these characters.

This changes the negation to only take effect in a wildcard (`*`) has
been seen somewhere in the expression, so in order to delete `-foo1` and
`-foo2` the following now works:

```
DELETE /-foo*
```

As well as:

```
DELETE /-foo1,-foo2
```

so in order to actually delete everything except for the "foo" indices
(ie, `test1` and `test2`) a user would now issue:

```
DELETE /*,--foo*
```

Relates to #19800

### CODE REVIEW

**Comment by nik9000** [comment]
Wheh. Pure negatives getting an implicit `*` before them very much the way we do things but it is so dangerous for stuff like this....

**Comment by nik9000** [comment]
So, yeah, I think this is a good idea. Though I expect it needs docs updates.

**Comment by nik9000** [comment]
I wonder if it'd be more clear to just reject pure negative index patterns entirely.....

**Comment by javanna** [comment]
> I wonder if it'd be more clear to just reject pure negative index patterns entirely.....

I tend to agree, it is confusing to accept `-foo*` but we need to as long as there can be existing indices with name starting with `-`, otherwise you can't refer to them anymore. We can reject such requests once we are sure that such indices are not there anymore, and that will be from 6.0 on I believe (as such indices can only come from 2.x, which will need reindexing to go to 6.0).

**Comment by dakrone** [comment]
> We can reject such requests once we are sure that such indices are not there anymore, and that will be from 6.0 on

I can open up a followup after this for rejecting these in the master branch only, once we get this one hammered out.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 1]
```diff
@@ -634,6 +636,10 @@ boolean isPreserveAliases() {
                 if (!noIndicesAllowedOrMatches(options, matches)) {
                     throw infe(expression);
                 }
+
+                if (expression.contains("*")) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: what happens with `+test1, +test2, +test3,-test2` ?  it looks silly but should be supported and resolve to test1 and test3 only. I think not only a wildcard expression can be before a negation, but there has to be something, anything that is not a negation?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: At this point I wonder if the order should still matter as it used to. Maybe negations should just subtract from the rest of the indices mentioned in the expression, with the requirement that an expression cannot be a negation only, otherwise we interpret it as an explicit index name that contains `-`. Should `-test1,test*` resolve to the same indices as `test*,-test1`? What other edge cases am I missing here? I also wonder why we support `+` as it is implicit anyway when missing.

**Review by javanna — CHANGES_REQUESTED** [review]
left a couple of comments, I think we may want to update api-conventions.asciidoc also, and look if we have this negation anywhere in our docs, not sure.

**Comment by javanna** [comment]
> I can open up a followup after this for rejecting these in the master branch only, once we get this one hammered out.

sounds great thanks

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 1 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: > what happens with `+test1, +test2, +test3,-test2`

That would resolve to: `["test1", "test2", "test3", "-test2"]`, assuming a `-test2` index exists, which I think is the correct behavior?

If `-test2` doesn't exist then it won't be negated and in that case I'm not sure what we want to do, do you think we should support this (adding and then removing an index without wildcards)?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: > `+` as it is implicit anyway when missing

+1 to remove `+` as an operator in a followup for 6.0-only, it's just confusing

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: > That would resolve to: ["test1", "test2", "test3", "-test2"], assuming a -test2 index exists, which I think is the correct behavior?

I see, yea that's ok, just different compared to before (we would previously add and remove `test2`, but there was no way to refer to `-test2` if existing I believe), but correct. Do we have a test for this too?

I think documenting all this under the api conventions page would clear things up.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: How about the ordering thing that I mentioned before? Should `-test1,test*` resolve to the same indices as `test*,-test1`? It doesn't now right?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: > Should `-test1,test*` resolve to the same indices as `test*,-test1`? It doesn't now right?

It doesn't right now, I think it would depend on if a `-test1` index existed.

I don't think it should though, since the intended behavior would be different regardless if you swapped the ordering of negations? What do you think?

Maybe instead we should start with a list of indices and work backwards towards what we expect for certain combinations?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Also `-test1,*test2*,-test20` or something along those lines? :)

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: Sure, I've added both of those tests :)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 2]
```diff
@@ -579,6 +579,7 @@ boolean isPreserveAliases() {
 
         private Set<String> innerResolve(Context context, List<String> expressions, IndicesOptions options, MetaData metaData) {
             Set<String> result = null;
+            boolean wildcardSeen = false;
             for (int i = 0; i < expressions.size(); i++) {
                 String expression = expressions.get(i);
                 if (aliasOrIndexExists(metaData, expression)) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: do you understand this if block? I suspect it made sense before your change, to make it possible to refer to existing indices or aliases that started with `-` rather than treating the name as a negation. That said, I can't quite follow why it makes sense in some cases for `result` to be `null`. This was already there, not your doing but I wonder if it's related and may be cleaned up.

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: can we use the existing `Regex.isSimpleMatchPattern` instead?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 3]
```diff
@@ -634,6 +636,10 @@ boolean isPreserveAliases() {
                 if (!noIndicesAllowedOrMatches(options, matches)) {
                     throw infe(expression);
                 }
+
+                if (expression.contains("*")) {
+                    wildcardSeen = true;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I wonder if we should set this to true only if options.expandWildcardsOpen || options.expandWildcardsClosed . Otherwise wildcard expressions don't get expanded...

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 2 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I don't know why it doesn't initialize the `result` set like the other null checks do, but I do think *all* of this code should be cleaned up after this (I hesitate to touch more code as this is supposed to be backported all the way back to 2.4)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java (hunk 3 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I agree, I've changed the `if` to check these

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 4]
```diff
@@ -637,7 +637,8 @@ boolean isPreserveAliases() {
                     throw infe(expression);
                 }
 
-                if (expression.contains("*")) {
+                if (Regex.isSimpleMatchPattern(expression) &&
+                        (options.expandWildcardsOpen() || options.expandWildcardsClosed())) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: add some tests around checking the options?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java]: I actually reverted the change to add the check, because if wildcards are both disabled then this check is already handled at a higher level, in `resolve` instead of `innerResolve`:

https://github.com/elastic/elasticsearch/blob/e57720e09170208e8059c39d43b848024d79ed9d/core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L559-L561

**Review by javanna — APPROVED** [review]
